### DISCUSSION TITLE
values/errors read

### DISCUSSION THREAD

**Discussion by archiron** [scikit-hep/uproot5#466] [General]
Hi, I started to work with uproot and I have a weird effect;
if after opening a root file and reading an histo [1], I try to read the _values/_errors arrays [2], I only empty arrays returned.

If before reading I use a statement such as : 
        values_1 = hist_1_4.values()
        errors_1 = hist_1_4.errors()
then, I can have access to the _values/_errors arrays as in point [2].
Is this way correct and known ?

Arnaud

[1] : fileName1 = rootFilesPath + '/' + "DQM_V0001_R000000001__RelValZEE_14__CMSSW_12_1_0_pre2-121X_mcRun3_2021_realistic_v1-v1__DQMIO.root"
    histoPath = "DQMData/Run 1/EgammaV/Run summary/ElectronMcSignalValidator"
    datasets = ['h_ele_PhiMnPhiTrueVsEta_pfx']
    rootFile = up4.open(fileName1)
    root_tree_1_4 = rootFile[histoPath]
    hist_1_4 = root_tree_1_4[dataset]

[2] : for n1, v1 in hist_1_4.__dict__.items():
            if n1.startswith("_values"):
                allValues_1 = v1
            elif n1.startswith("_errors"):
                allErrors_1 = v1

**Comment by jpivarski** [comment]
I don't remember what's in `_values` and `_errors`, but they're probably caches.

A general rule in Python is that anything that starts with an underscore is an internal detail, equivalent to `private` in C++, not part of the library's interface. Private functions/classes/attributes/variables won't, in general, behave in predictable ways, and they can be removed or changed from one version of a library to the next.

So the important question is whether `values()` and `errors()` behave as you expect. These methods don't start with underscores and they're documented (here: [values](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TH1.TH1.html#values) and [errors](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TH1.TH1.html#errors), and also as satisfying a [general protocol (UHI)](https://uhi.readthedocs.io/en/latest/plotting.html#using-the-protocol), which gives the same meaning to these methods across multiple libraries). Thus, they're part of the public interface.

Also, it would help if code snippets in GitHub are [surrounded by triple-backticks](https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks), since this formats them as code. Particularly in Python, the meaning of the code depends on indentation, and without the formatting, the indentation is removed!

**Reply by archiron** [reply]
Hi, thanks for your reply.
I actually wanted to use `_values `/ `_errors` because It does not contains under/overflow values. But in fact, this is easier to use _official_ form such as `values()` or `errors()`.

Arnaud

P.S.
sorry for the code format I forgot to use.
